src/cpp/core/include/core/Thread.hpp

```diff
@@ -264,17 +264,55 @@ class ThreadsafeQueue : boost::noncopyable
 
    bool deque(T* pVal, const boost::posix_time::time_duration& waitDuration)
    {
-      // first see if we already have one
-      if (deque(pVal))
+      boost::unique_lock<boost::mutex> lock(*pMutex_);
+
+      if (!queue_.empty())
+      {
+         *pVal = queue_.front();
+         queue_.pop();
+
          return true;
+      }
+      else if (waitDuration.is_not_a_date_time())
+      {
+         while (queue_.empty()) 
+            pWaitCondition_->wait(lock);
```

Comment:
I wonder if we should check for / catch exceptions as we do below? https://www.boost.org/doc/libs/1_87_0/libs/fiber/doc/html/fiber/synchronization/conditions.html#condition_variable_wait suggests that `wait` can potentially throw.